CLI tool for managing JupyterLab extensions

cronan03 opened the pull request:
This pull request introduces a CLI (Command Line Interface) tool for managing JupyterLab extensions, designed to enhance developer workflow. The CLI tool, named jupyterlab-builder, provides commands for developing, building, and watching JupyterLab extensions. It leverages the traitlets library for configuration and flexibility.

**Key Features**:

Develop Command: Allows developers to set up a development environment for lab extensions, facilitating iterative development and testing.
Build Command: Provides a command for building lab extensions, enabling packaging and distribution.
Watch Command: Enables developers to watch lab extensions for changes and automatically rebuild them, improving productivity during development.
BaseExtensionApp: Implements the core logic for parsing commands and delegating them to appropriate subcommands (DevelopLabExtensionApp, BuildLabExtensionApp, WatchLabExtensionApp).
Usage of Traitlets: Utilizes the traitlets library for defining configuration options and managing application state, ensuring flexibility and ease of customization.

fcollonval reviewed (commented):
This is going in the right direction @cronan03 

I added some inline comments. And could you rebase on the main branch as I setup a skeleton structure for the package in https://github.com/jupyterlab/jupyterlab-builder/pull/2?

cronan03 commented on base_extension_app.py:
```diff
@@ -0,0 +1,94 @@
+import os
+from traitlets import Bool, Unicode, default, List, Instance
+from jupyterlab.labapp import LabApp
+from jupyterlab.commands import build
```
So as I understand there should only be an __init__.py file along with a separate file for each functionality extracted like build.py, watch.py etc.  All other extracted logic , like contents of these above mentioned imports should be a part of the __init__.py file. Am I correct?

fcollonval replied:
You can keep the `base_extension_app.py` and `federated_extensions.py` files. You can also create (or not) a dedicated file to bring the code from those imports:

```py
from jupyterlab.labapp import LabApp
from jupyterlab.commands import build
```

For the `LabApp` instead of copying all the code (that should be mostly useless), it will be nice to start from an empty class and run the extracted code to see what is actually needed.

My guess is `LabApp` inherits from [`LabServerApp`](https://github.com/jupyterlab/jupyterlab_server/blob/cca5a50acfedfb45a843acf733ddbb410634451e/jupyterlab_server/app.py#L18) that itself inherits of [`LabConfig`](https://github.com/jupyterlab/jupyterlab_server/blob/cca5a50acfedfb45a843acf733ddbb410634451e/jupyterlab_server/config.py#L219). And only attributes of that `LabConfig` class should be needed here.

cronan03 commented:
Hi @fcollonval I have rebased my code on your #2 commit. I have also extracted the methods and classes being used in the federated_extensions.py file. ie
from jupyter_core.paths import ENV_JUPYTER_PATH, SYSTEM_JUPYTER_PATH, jupyter_data_dir
from jupyter_core.utils import ensure_dir_exists
from jupyter_server.extension.serverextension import ArgumentConflict
from jupyterlab_server.config import get_federated_extensions

I will now move on to extracting code used in these imports
from jupyterlab.labapp import LabApp
from jupyterlab.commands import build as per your comments above.

Please let me know if I did this correctly.

fcollonval commented:
Thanks a lot @cronan03 

I realize my review comments was not specific enough. It is ok to depend on `jupyter_core` and `jupyter_server`. The only we should not depend on are `jupyterlab` and `jupyterlab_server`. Sorry about that as it made you copy to much code in this PR.

fcollonval commented:
If you want to check the imported code, you can also import here the tests for that part of the code. They are located in the file: https://github.com/jupyterlab/jupyterlab/blob/6152e3a5f94d728ccd2727d6d8413f12ddafb669/jupyterlab/tests/test_jupyterlab.py#L156
